x/session/sessionNodes.go

```diff
@@ -0,0 +1,141 @@
+package session
+
+import (
+	"encoding/hex"
+	"github.com/pokt-network/pocket-core/crypto"
+	"github.com/pokt-network/pocket-core/tests/fixtures"
+	"github.com/pokt-network/pocket-core/types"
+	"sort"
+)
+
+// A simple slice abstraction of type `Node`
+// All of the active nodes on the Pocket Network
+type AllActiveNodes []types.Node
+
+// The computational distance between two byte arrays
+// this is judged by XORing the two
+type ComputationalDistance []byte
+
+// A simple slice abstraction of type `Node`
+// These nodes are linked to the session
+type SessionNodes []types.Node
+
+// A node linked to it's computational distance
+type NodeDistance struct {
+	types.Node
+	distance ComputationalDistance
+}
+
+type NodeDistances []NodeDistance
+
+func NewSessionNodes(nonNativeChain SessionBlockchain, sessionKey SessionKey) (SessionNodes, error) {
+	if len(sessionKey) == 0 {
+		return nil, EmptySessionKeyError
+	}
+	if len(nonNativeChain) == 0 {
+		return nil, EmptyNonNativeChainError
+	}
+	// using mock world state
+	allActiveNodes, err := fixtures.GetNodes()
+	if err != nil {
+		return nil, err
+	}
+	// session nodes are just a wrapper around node slice
+	var result SessionNodes
+	// node distance is just a node with a computational field attached to it
+	var xorResult []NodeDistance
+	// ensure there is atleast the minimum amount of nodes
+	if len(*allActiveNodes) < SESSIONNODECOUNT {
+		return nil, InsufficientNodesError
+	}
+	// filter `allActiveNodes` by the HASH(nonNativeChain)
+	result, err = filter(AllActiveNodes(*allActiveNodes), nonNativeChain)
+	if err != nil {
+		return nil, err
+	}
+	// xor each node's public key and session key
+	// return NodeDistance array to be ordered
+	xorResult, err = xor(result, sessionKey)
+	if err != nil {
+		return nil, err
+	}
+	// sort the nodes based off of distance
+	result = revSort(xorResult)
+
+	// return the top 5 nodes
+	return result[0:5], nil
+}
+
+// filter the nodes by non native chain
+func filter(allActiveNodes AllActiveNodes, nonNativeChainHash []byte) (SessionNodes, error) {
+	var result SessionNodes
+	for _, node := range allActiveNodes {
+		if !node.SupportedChains.Contains(hex.EncodeToString(nonNativeChainHash)) {
+			continue
+		}
+		result = append(result, node)
+	}
+	if len(result) == 0 {
+		return nil, InsufficientNodesError
+	}
+	return result, nil
+}
+
+// xor the sessionNodes.publicKey against the sessionKey to find the computationally
+// closest nodes
+func xor(sessionNodes SessionNodes, sessionkey SessionKey) (NodeDistances, error) {
```

Comment:
Shouldn't it be better for randomness to xor the address? Since it's a hash of the public key?

Reply:
I re-hash the public key to ensure that it's the same length as the sessionKey

Reply:
Got it, that makes sense.